src/contracts/GPv2Settlement.sol

```diff
@@ -288,6 +293,11 @@ contract GPv2Settlement {
             );
         }
 
+        require(trade.feeDiscount <= BPS_BASE, "GPv2: invalid fee discount");
+        executedFeeAmount =
+            executedFeeAmount.mul(BPS_BASE - trade.feeDiscount) /
```

Comment:
I don't think it matters much but giving the discount in basis points (of already something that is in basis points) allows an extremely (likely overly) fine granular control of how much fee we want to charge (basically parts per million). I think having one byte fee discount (and treating it from 1-100) would be reasonable as well. 
Maybe keeping everything the same unit (basis points) is more coherent though?

Reply:
For the typical order with a discount of 0, this would amount to saving 4 gas in the calldata. For discounts greater than 2.5% (255 bps) it would be represented in 2 bytes meaning 16 gas saving in the calldata. The math in the contract itself is unaffected. On thing we can do to reduce the gas cost in the case we want to do a full discount is change this to
```solidity
if (trade.feeDiscount > BPS_BASE) {
  executedFeeAmount = 0;
} else {
  executedFeeAmount = executedFeeAmount.mul(BPS_BASE - trade.feeDiscount) / BPS_BASE;
}
```

This would make it cost only 4 more gas to do a full refund than if we only had 1 byte.

Reply:
I think the gas cost of doing 1 vs 2 bytes is quite negligible since we only need to pay for the extra calldata which is relatively  cheap.

Reply:
Doesn't the require add an extra cost for allocating the string, that is the reason why we avoid `.div`?
Otherwise I'm not really concerned about the gas consumption.

On the other hand, I agree that the amount of precision is "weird". It's not granular enough that allows us to choose the fee the user pays ("only pay 1 DAI for every order between $NEWTOKEN and DAI!") but it's also more than needed for a coarse usage ("fees down by 30% when trading $NEWTOKEN!"). I'd be in favor of reducing it to 1 byte. In any case I don't think it matters too much.

Reply:
> Doesn't the require add an extra cost for allocating the string, that is the reason why we avoid .div?

No. The extra string allocation is because `div` is implemented as:
```sol
function div(uint256 a, uint256 b) internal pure returns (uint256);
function div(uint256 a, uint256 b, string memory reason) internal pure returns (uint256);
```

So the string gets allocated to be passed into the function call. Essentially, the code does this:
```sol
string memory reason = "default reason";
// div(a, b, reason) {
  if (b == 0) {
    revert(reason);
  }
// }
```

In the case of `require(condition, reason)`, the string literal only gets written to memory for the `REQUIRE` opcode to read from if the condition fails. In other words, it generates code closer to:
```sol
if (!condition) {
  string memory reason = "...";
  revert(reason);
}
```

Reply:
I created https://github.com/gnosis/gp-v2-contracts/issues/252 to discuss what unit to use for the fee discount.